### PR TITLE
ColumnTable PutInto & DeleteFrom API

### PR DESCRIPTION
**Pull request by rishitesh**
## Changes proposed in this pull request

Column table putInto operation: This PR is the first stage of implementation , where two different jobs are executed to achieve the goal. One is update & another is insert.
Both the operation depends on the inbuilt insert & update operator. The difference is the where clause where each one has a different join query. 
This is an the first step for column table putInto. This approach will cause inconsistent data if concurrent inserts are going on.
As in some of the use cases streaming is the only way to insert data, this will work fine for them.

Changes 
a) Implemented a new logical plan PutIntoColumnTable which is a wrapper over insert & update. Insert has an inner join as child , where as Update has an antijoin as child.
b) Column table can now accept "key_columns" as an optional parameter. This parameter is stored in HiveStore catalog.
c) The join condition for insert and update is derived out of key columns.
d) Bypassed PreWriteCheck at Spark layer for column table putInto, But as mentioned earlier this needs to be changed once we need to support concurrent update.
e) Added support for df.deleteFrom() api for column tables. This was relatively straight forward as column tables already supported delete with a subquery in SQL. deleteFrom API will also rely on key_columns to match records. 

Also fixed some issues with Update/Delete statement 
**delete from col_table a where exists (select 1 from row_table b where a.col2 = b.col2 and a.col3 = b.col3)**
Parser was expecting only table identifier and was failing for any alias. 

**update col_table set a.col3 = b.col3, a.col4 = b.col4 from col_table a join row_table b on a.col2 = b.col2**
**update col_table set a.col3 = b.col3, a.col4 = b.col4 from col_table a, row_table b where a.col2 = b.col2**
Supported the above syntaxes for co-related update 

## Patch testing
pre-checkin pending

## ReleaseNotes.txt changes

NA

## Other PRs 

https://github.com/SnappyDataInc/snappy-aqp/pull/162

### CODE REVIEW

**Comment by kneeraj** [comment]
>> This approach will cause inconsistent data if concurrent inserts are going on.
Do we have a ticket for this. If not please file one. And also do we have a rough idea as to how we would handle concurrent operations.

**Comment by rishitesh** [comment]
@kneeraj . Yes for concurrent inserts on same key will make it inconsistent. This API will be an experimental features for column tables where ingestion is only from a single source like streaming. The long term plan would be to merge both insert & update into a single plan. That would ensure per partition consistency.

**Comment by sumwale** [comment]
@rishitesh Just merging the plans won't be enough to fix the concurrency problem. Essential design problem is that snapshot isolation itself does not work for constraint checking + inserts. In particular two threads can determine from their own snapshots that entries have to be inserted so both go ahead and insert on the same key values.

One solution to handle that is to detect a read-write conflict for such a case but that will make it completely serial and not concurrent because there will almost always be a read-write conflict as soon as two threads have inserts even if those are not on same keys.

Another solution is to do the inserts into row buffer which uses the key columns as PK and force it to not rollover as long as there are some open transactions but it has huge performance problems as we know from past (e.g. on my laptop row buffer inserts give max 8K-10K insert rate compared to 1M or more directly). Alternatively instead of using row buffer that has a high overhead, use a temporary in-memory hash map that holds the values inserted in each transaction at the time of its commit that also includes the batchId+ordinal+RVV used in the batch (locking map for duration of commit). The second thread will lookup this map and change to delta update using the matched batchId+ordinal at the last step if matching rows are found. The idea being to minimize the lock duration to increase concurrency as much as possible. Cleanup of this map will be done using the RVV if its determined that no active TX has a snapshot on that RVV.

Once the indexing work is in, then another solution is to merge the latest value (i.e. last one wins) instead of adding to it because both will go into the same delta buffer and thus the region entry. However, need to somehow ensure there that all column deltas from one or other win and not that some columns for one thread win but others from other thread. This can be done by having a consistent policy on resolution of who will win e.g. second thread must ensure it gets all entry versions as bigger ones than the first (or vice versa) and then the merge will always prefer the larger entry version. Also like for row buffer approach this needs to ensure the delta merge into higher delta or base column should not happen while there are open transactions that have taken a snapshot that includes the delta.

**Comment by rishitesh** [comment]
Merging for regression. Will incorporate review comments later.

File: cluster/src/test/scala/org/apache/spark/sql/execution/benchmark/ColumnCacheBenchmark.scala [hunk 1]
```diff
@@ -102,13 +106,57 @@ class ColumnCacheBenchmark extends SnappyFunSuite {
...
+    def cleanup(): Unit = {
+      snappySession.sql("drop table if exists test")
+      doGC()
+    }
+
+    def testCleanup(): Unit = {
+      snappySession.sql("truncate table if exists test")
+    }
+    // As expected putInto is two times slower than a simple insert
+    addCaseWithCleanup(benchmark, "Insert", numIters, prepare, cleanup, testCleanup) { i =>
```
[the remark this answers is not in the document]
Comment (rishitesh) [reply cluster/src/test/scala/org/apache/spark/sql/execution/benchmark/ColumnCacheBenchmark.scala]: Yes. Was looking how to do it on a plan level. For example if we can send a boolean flag that all rows are consumed, then further operations will be saved.

File: core/src/main/scala/io/snappydata/Literals.scala [hunk 2]
```diff
@@ -328,6 +328,11 @@ object Property extends Enumeration {
     s"Boolean if false tells engine to use bootstrap analysis for error calculation for all cases" +
       s". Default is true.", Some(true), null, false)
 
+  val PutIntoInnerJoinCacheSize =
+    SQLVal[Long](s"${Constant.PROPERTY_PREFIX}cache.putIntoInnerJoinResultSize",
+    "The putInto inner join would be cached if the table is of size less " +
+        "than PutIntoInnerJoinCacheSize. Default value is 100 MB.", Some(100L * 1024 * 1024))
```
[the remark this answers is not in the document]
Comment (rishitesh) [reply core/src/main/scala/io/snappydata/Literals.scala]: Correct, will change it.

**Pull request merged**